pkg/build/graph/edges.go

```diff
@@ -37,48 +56,68 @@ func AddAllBuildEdges(g osgraph.MutableUniqueGraph) {
 	}
 }
 
-// AddInputOutputEdges links the build config to other nodes for the images and source repositories it depends on.
-func AddInputOutputEdges(g osgraph.MutableUniqueGraph, node *buildgraph.BuildConfigNode) *buildgraph.BuildConfigNode {
-	output := node.BuildConfig.Spec.Output
-	to := output.To
-	switch {
-	case to == nil:
-	case to.Kind == "DockerImage":
-		out := imagegraph.EnsureDockerRepositoryNode(g, to.Name, "")
-		g.AddEdge(node, out, BuildOutputEdgeKind)
-	case to.Kind == "ImageStreamTag":
-		out := imagegraph.FindOrCreateSyntheticImageStreamTagNode(g, imagegraph.MakeImageStreamTagObjectMeta2(defaultNamespace(to.Namespace, node.BuildConfig.Namespace), to.Name))
-		g.AddEdge(node, out, BuildOutputEdgeKind)
+func imageRefNode(g osgraph.MutableUniqueGraph, ref *kapi.ObjectReference, bc *buildapi.BuildConfig) graph.Node {
+	if ref == nil {
+		return nil
 	}
+	switch ref.Kind {
+	case "DockerImage":
+		if ref, err := imageapi.ParseDockerImageReference(ref.Name); err == nil {
+			tag := ref.Tag
+			ref.Tag = ""
+			return imagegraph.EnsureDockerRepositoryNode(g, ref.String(), tag)
+		}
+	case "ImageStream":
+		return imagegraph.FindOrCreateSyntheticImageStreamTagNode(g, imagegraph.MakeImageStreamTagObjectMeta(defaultNamespace(ref.Namespace, bc.Namespace), ref.Name, imageapi.DefaultImageTag))
+	case "ImageStreamTag":
+		return imagegraph.FindOrCreateSyntheticImageStreamTagNode(g, imagegraph.MakeImageStreamTagObjectMeta2(defaultNamespace(ref.Namespace, bc.Namespace), ref.Name))
+	case "ImageStreamImage":
+		return imagegraph.FindOrCreateSyntheticImageStreamImageNode(g, imagegraph.MakeImageStreamImageObjectMeta(defaultNamespace(ref.Namespace, bc.Namespace), ref.Name))
+	}
+	return nil
+}
+
+// AddOutputEdges links the build config to its output image node.
+func AddOutputEdges(g osgraph.MutableUniqueGraph, node *buildgraph.BuildConfigNode) {
+	out := imageRefNode(g, node.BuildConfig.Spec.Output.To, node.BuildConfig)
+	g.AddEdge(node, out, BuildOutputEdgeKind)
+}
 
+// AddInputEdges links the build config to its input image and source nodes.
+func AddInputEdges(g osgraph.MutableUniqueGraph, node *buildgraph.BuildConfigNode) {
 	if in := buildgraph.EnsureSourceRepositoryNode(g, node.BuildConfig.Spec.Source); in != nil {
 		g.AddEdge(in, node, BuildInputEdgeKind)
 	}
+	inputImage := buildutil.GetImageStreamForStrategy(node.BuildConfig.Spec.Strategy)
+	if input := imageRefNode(g, inputImage, node.BuildConfig); input != nil {
+		g.AddEdge(input, node, BuildInputImageEdgeKind)
```

Comment:
input could end up being nil here (docker type build w/ no From defined).  is that a problem?